lib/rucio/core/did.py

```diff
@@ -3002,3 +3002,53 @@ def insert_deleted_dids(filter_: "ColumnExpressionArgument[bool]", *, session: "
             is_archive=did.is_archive,
             constituent=did.constituent
         ).save(session=session, flush=False)
+
+@transactional_session
+def add_files_with_structure(
```

Comment:
I think the method name could be a bit clearer - initially I thought that `structure` meant "file structure", maybe referring to the pfn, bytes, adler32, but after reading the method I understood it means that it adds a container DID for each parent path. Maybe something like `add_files_and_container_dids`